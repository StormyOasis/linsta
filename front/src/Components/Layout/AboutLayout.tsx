import React, { useEffect, useState } from "react";
import { keyframes, styled } from "styled-components";
import StyledButton from "../Common/StyledButton";
import { Div, Flex, FlexRowFullWidth, Link } from "../Common/CombinedStyling";
import {
    FaReact,
    FaDocker,
    FaGithub,
    FaAws,
    FaHtml5,
    FaCss3Alt,
    FaNodeJs,
    FaJs,
    FaJira,    
} from 'react-icons/fa';

import {
    SiRedux,
    SiStyledcomponents,
    SiTypescript,
    SiJest,
    SiKoa,
    SiElastic,
    SiServerless,
    SiRedis,
    SiJsonwebtokens,
    SiNginx,
    SiGrafana,
    SiKibana,
    SiInsomnia,
    SiFacebook,
    SiGraphql,
    SiAxios,    
} from 'react-icons/si';
import { SiLocation } from 'react-icons/si';
import { SiAmazonaws
} from 'react-icons/si';
import LargeLogo from "../Common/LargeLogo";
import StyledLink from "../Common/StyledLink";


const fadeInUp = keyframes`
    from {
        opacity: 0;
        transform: translateY(20px);
    }
    to {
        opacity: 1;
        transform: translateY(0);
    }
`;

const fadeIn = keyframes`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
`;

const Wrapper = styled(Div)`
    padding: 2rem 1rem;
    max-width: 1100px;
    margin: 0 auto;
    color: #333;
    line-height: 1.6;
`;

const Section = styled.section`
    margin-bottom: 3rem;
    animation: ${fadeInUp} 0.8s ease both;
`;

const Subtitle = styled.h2`
    font-size: 1.75rem;
    color: #222;
    margin-bottom: 1rem;
`;

const Text = styled.p`
    font-size: 1rem;
    margin-top: 0.5rem;
`;

const StackWrapper = styled(Flex)`
    flex-wrap: wrap;
    gap: 2rem;
    margin-top: 1rem;

    @media (max-width: ${props => props.theme["breakpoints"].md}px) {
        flex-direction: column;
    }
`;

const StackColumn = styled(Div)`
    flex: 1;
    min-width: 250px;
`;

const StackHeading = styled.h3`
    font-size: 1.2rem;
    color: #444;
    margin-bottom: 0.5rem;
`;

const TechList = styled.ul`
    list-style: none;
    padding: 0;
    display: flex;
    flex-direction: column;
    gap: 0.5rem;

    li {
        background: #f1f1f1;
        padding: 0.5rem 1rem;
        border-radius: 4px;
        transition: background 0.3s ease;

        &:hover {
            background: #e0e0e0;
        }
    }
`;

const JiraList = styled(TechList)`    
    padding-left: 48px;
    padding-right:48px;
`;

const ScreenshotGrid = styled(Div)`
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
    gap: 1.5rem;
`;

const ScreenshotCard = styled(Div)`
    position: relative;
    overflow: hidden;
    border-radius: 6px;
    border: 1px solid #ddd;
    box-shadow: 0 2px 6px rgba(0,0,0,0.1);
    cursor: pointer;

    &:hover div {
        opacity: 1;
        transform: translateY(0);
    }
`;

const ScreenshotImg = styled.img`
    width: 100%;
    display: block;
    height: auto;
`;

const ScreenshotOverlay = styled(Div)`
    position: absolute;
    bottom: 0;
    background: rgba(0,0,0,0.7);
    color: #fff;
    width: 100%;
    padding: 1rem;
    opacity: 0;
    transform: translateY(100%);
    transition: all 0.3s ease;

    p {
        margin: 0;
        font-size: 0.95rem;
        line-height: 1.4;
    }

    @media (hover: none) and (pointer: coarse) {
        opacity: 1;
        transform: translateY(0);
        position: static;
        background: #333;
        color: #fff;
        margin-top: 0.5rem;
    }
`;


const Footer = styled.footer`
    text-align: center;
    font-size: 0.9rem;
    margin-top: 3rem;
`;

const WarningText = styled(Div)`
    text-align: center;    
    font-weight: 700;
    font-size: .99em;
`;

const VisitWrapperLink = styled(Link)`
    display: flex;
    justify-content: center;
`;

const VisitButton = styled(StyledButton)`
    margin: 0;
    @media (max-width: ${props => props.theme["breakpoints"].md}px) {
        text-align: center;
        width: 100%;
    }
`;

const LogoWrapper = styled(Div)`
    justify-self: center;
`;

const ModalBackdrop = styled.div`
    position: fixed;
    inset: 0;
    background: rgba(0,0,0,0.7);
    display: flex;
    justify-content: center;
    align-items: start;
    z-index: 1000;
    animation: ${fadeIn} 0.3s ease forwards;
`;

const ModalContent = styled.div`
    margin-top: 24px;
    background: #fff;
    padding: 1rem;
    border-radius: 8px;
    max-width: 75%;
    max-height: 75%;
    overflow: auto;
    position: relative;
    box-shadow: 0 0 10px rgba(0,0,0,0.3);

    animation: ${fadeIn} 0.3s ease forwards;

    img {
        width: 100%;
        height: auto;
        display: block;
        border-radius: 6px;
        margin-bottom: 1rem;
    }

    p {
        font-size: 0.95rem;
        color: #333;
    }
`;

const CloseButton = styled.button`
    position: absolute;
    top: 0.5rem;
    right: 0.75rem;
    font-size: 1.5rem;
    background: none;
    border: none;
    color: #333;
    cursor: pointer;

    &:hover {
        color: #000;
    }
`;

const imageUrl = `https://d1xxvwtswm7wsd.cloudfront.net/about`
const screenshots = [
    {
        src: `${imageUrl}/signup.png`,
        alt: 'Sign up form',
        description: 'User registration form with validation and backend integration via Koa + JWT',
    },
    {
        src: `${imageUrl}/confirmationEmail.png`,
        alt: 'Sign up confirmation',
        description: 'Confirmation email received as part of signup flow',
    },
    {
        src: `${imageUrl}/mainFeed.png`,
        alt: 'Main feed',
        description: "The user's primary feed, displaying posts from the specific users they follow"
    },
    {
        src: `${imageUrl}/exploreWithSearch.png`,
        alt: 'Explore with search',
        description: 'The Explore page showing posts and includes the search popout with example search results',
    },
    {
        src: `${imageUrl}/createPost.png`,
        alt: 'Create post',
        description: 'The final step in the create new post flow, showing adding captions, location, collaborators, and additional settings',
    },
    {
        src: `${imageUrl}/commentModal.png`,
        alt: 'Comment modal',
        description: 'Modal where a user can add, like, and reply to comments',
    },
    {
        src: `${imageUrl}/profilePage.png`,
        alt: 'Profile',
        description: "The profile page showing a user's bio and additional information, as well as showing each of that user's posts"
    },
    {
        src: `${imageUrl}/editProfile.png`,
        alt: 'Edit profile',
        description: "Page allowing the user to update their bio, profile picture, webpage, and additional information"
    },
    {
        src: `${imageUrl}/metrics.png`,
        alt: 'Metrics',
        description: "Example Grafana visualizations based on metrics collected via statsd"
    },
    {
        src: `${imageUrl}/createPostMobile.png`,
        alt: 'Create post on mobile',
        description: 'The final step in the create post flow as seen on mobile',
    },
    {
        src: `${imageUrl}/mainFeedMobile.png`,
        alt: 'Main feed on mobile',
        description: 'The main feed as it appears on mobile devices',
    },
];

const AboutLayout: React.FC = () => {
    const [modalImage, setModalImage] = useState(null);

    const openModal = (image: any) => setModalImage(image);
    const closeModal = () => setModalImage(null);

    useEffect(() => {
        const handleEsc = (e: any) => {
            if (e.key === 'Escape') closeModal();
        };
        window.addEventListener('keydown', handleEsc);
        return () => window.removeEventListener('keydown', handleEsc);
    }, []);

    return (
        <Wrapper>
            <Section>
                <LogoWrapper>
                    <Link href="/login" aria-label="Visit Linstagram">
                        <LargeLogo></LargeLogo>
                    </Link>
                </LogoWrapper>
                <Text style={{ textAlign: "center" }}>
                    A full-featured, enterprise-grade social networking platform inspired by Instagram. Built with a cloud-native architecture for high performance, scalability, and real-time interactivity.
                </Text>
                <WarningText>
                    **Note: This is not meant for public use nor is it in any way affilated with Meta.**
                </WarningText>
            </Section>

            <Section>
                <Subtitle>Project Overview</Subtitle>
                <Text>
                    This project replicates core social media functionality. Users can create profiles, post media, comment, like, and follow others. In addition, it includes enterprise-ready features such as secure authentication, AWS-powered geolocation suggestions, and a distributed event-driven backend architecture.
                </Text>
            </Section>

            <Section>
                <Subtitle>Tech Stack</Subtitle>
                <StackWrapper>
                    {/* Front-End */}
                    <StackColumn>
                        <StackHeading>Front-End</StackHeading>
                        <TechList>
                            <li><FaReact /> React + SSR</li>
                            <li><SiRedux /> Redux</li>
                            <li><SiTypescript /> TypeScript</li>
                            <li><FaJs /> JavaScript</li>
                            <li><FaHtml5 /> HTML</li>
                            <li><FaCss3Alt /> CSS</li>
                            <li><SiStyledcomponents /> styled-components</li>
                            <li><SiJest /> Jest</li>
                            <li><SiFacebook /> Facebook Lexical</li>
                        </TechList>
                    </StackColumn>

                    {/* Back-End */}
                    <StackColumn>
                        <StackHeading>Back-End</StackHeading>
                        <TechList>
                            <li><FaNodeJs /> Node.js</li>
                            <li><SiKoa /> Koa</li>
                            <li><SiJsonwebtokens /> JWT</li>
                            <li><SiRedis /> Redis</li>
                            <li><SiElastic /> ElasticSearch</li>
                            <li><SiGraphql /> JanusGraph</li>
                            <li><FaAws /> AWS S3, SES, SNS</li>
                            <li><FaAws /> AWS Location Services</li>
                            <li><FaAws /> AWS SQS</li>
                            <li><SiAxios /> Axios connection to OpenAI</li>
                        </TechList>
                    </StackColumn>

                    {/* DevOps & Tools */}
                    <StackColumn>
                        <StackHeading>Infrastructure, DevOps & Tools</StackHeading>
                        <TechList>
                            <li><FaDocker /> Docker / Docker Compose</li>
                            <li><FaGithub /> GitHub</li>
                            <li><SiServerless /> Serverless Framework</li>
                            <li><FaAws /> AWS EC2, Lambda, Cloudfront</li>
                            <li><SiNginx /> NGINX</li>
                            <li><FaJira /> Jira</li>
                            <li><SiGrafana /> Grafana</li>
                            <li><SiKibana /> Kibana</li>
                            <li><SiInsomnia /> Insomnia REST Client</li>
                        </TechList>
                    </StackColumn>
                </StackWrapper>
            </Section>

            <Section>
                <Subtitle>Screenshots</Subtitle>
                <ScreenshotGrid>
                    {screenshots.map((screenshot, index) => (
                        <ScreenshotCard key={index} onClick={() => openModal(screenshot)}>
                            <ScreenshotImg src={screenshot.src} alt={screenshot.alt} />
                            <ScreenshotOverlay>
                                <p>{screenshot.description}</p>
                            </ScreenshotOverlay>
                        </ScreenshotCard>
                    ))}
                </ScreenshotGrid>
                {modalImage && (
                    <ModalBackdrop onClick={closeModal}>
                        <ModalContent onClick={(e) => e.stopPropagation()}>
                            <CloseButton onClick={closeModal}>&times;</CloseButton>
                            <img src={modalImage.src} alt={modalImage.alt} />
                            <p>{modalImage.description}</p>
                        </ModalContent>
                    </ModalBackdrop>
                )}
            </Section>

            <Section>
                <Subtitle>Challenges & Solutions</Subtitle>
                <Text>The biggest challenge was infrastructure cost. Since this project is not intended to generate revenue, I had to design for cost-efficiency without sacrificing architecture quality.</Text>
                <Text>The most notable tradeoff was with the graph database. The original plan was to develop locally using Gremlin in JanusGraph and then deploy to AWS Neptune. However, even a minimal Neptune database proved to be too expensive.  The solution was to instead self host a JanusGraph install on an existing EC2 Instance.</Text>
                <Text>On the other hand, I tested deploying the back-end REST Api to AWS Lambda instead of EC2.  While it would certainly save money, AWS Lambda cold starts introduced unacceptable latency for user-facing APIs. In this case, a continuously running EC2 instance offered a better balance between performance and cost.</Text>
            </Section>

            <Section>
                <Subtitle>What's Next</Subtitle>
                <Text>The following are some of the current pending high priority Jira tickets for expanding functionality:</Text>
                <JiraList>
                    <li><b>LINSTA-130</b>: Epic - Create notification and messaging systems</li>
                    <li><b>LINSTA-127</b>: Collect front end metrics and send to SQS / Lambdas for processing</li>
                    <li><b>LINSTA-131</b>: Adjust Gremlin to not use read then write transactions in order to prevent lock contention</li>
                    <li><b>LINSTA-133</b>: Setup Jenkins or AWS CodeBuild / Deploy for CI/CD</li>
                </JiraList>
            </Section>

            <Section>
                <VisitWrapperLink href="/login" aria-label="Visit Linstagram">
                    <VisitButton text="Visit Linstagram" />
                </VisitWrapperLink>
                <Div>
                    <Text><b>Note</b>: I’m currently operating in the AWS test (sandbox) environment for email and SMS, which restricts messaging to pre-approved addresses only.  As a result, the signup process will be blocked at the confirmation code step for now.</Text>
                    <Text>I’ve submitted a request to move to production mode. Once approved, I’ll be able to send messages to any address. </Text>
                    <Text>In the meantime, I have provided two valid logins for use in the app:</Text>
                    <FlexRowFullWidth>
                        <TechList>
                            <li>Username: linstatest01</li>
                            <li>Password: Linstatest01!</li>
                        </TechList>
                        <TechList style={{ marginLeft: "28px" }}>
                            <li>Username: linstatest02</li>
                            <li>Password: Linstatest02!</li>
                        </TechList>
                    </FlexRowFullWidth>
                </Div>
            </Section>

            <Footer>
                <StyledLink to="https://github.com/StormyOasis/linsta">
                    View Source on GitHub
                </StyledLink>
            </Footer>
        </Wrapper>
    );
};

export default AboutLayout;