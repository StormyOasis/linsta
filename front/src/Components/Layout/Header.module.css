.fixedHeader {
    position: fixed; 
    top: 0; 
    width: 100%;
    display: flex;
    flex-direction: column;
    border-bottom: 1px solid #EFEFEF;
    background-color: white;
}

.innerHeader {
    align-items: center;
    position: relative;
    display: flex;
    flex-direction: row;
    justify-content: center;
    padding: 0px;
    margin: auto; 
    width: 65%;
    height: 60px;
}

.actionWrapperInner {
    position: relative; 
    display: flex; 
    padding: 0; 
    padding-left: 24px;
    align-items: stretch;
}