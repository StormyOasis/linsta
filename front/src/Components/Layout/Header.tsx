import React from "react";
import { styled } from "styled-components";
import { Link } from "react-router-dom";

import LogoSVG from '/public/images/linsta_small.svg';
import SearchSVG from '/public/images/search.svg';

import * as styles from './Header.module.css';

const HeaderWrapper = styled.div`
  z-index: 100;
  align-items: center;
  flex-direction: column;
  display:flex;
  height: 60px;  
  width: 100%;
  padding: 0;
  margin: 0;
  position: relative;
  top: 0;  
`;

const LogoWrapper = styled.div`
  flex-basis: 128px;
  flex-grow: 1;
`;

const LogoImage = styled(LogoSVG)`
   display:flex;
   overflow: visible;
   background: transparent;
`;

const SearchImage = styled(SearchSVG)`
  overflow: hidden;
  height: 16px;
  line-height: 18px;
  width: 16px;
  position: relative;
  cursor: pointer;
  color: rgb(142, 142, 142);
  display: block;
`;

const SearchBox = styled.input`
  display:flex;
  padding-top: 3px;
  padding-bottom: 3px;
  padding-left: 16px;
  padding-right: 16px;
  border-radius: 8px;
  border: none;
  font-size: 16px;
  height: 36px;
  line-height: 18px;
  text-align: start;
  width: 350px;
  cursor: pointer;
  background-color: ${props => props.theme["input"].backgroundColor};
`;

const SearchWrapper = styled.div`
  align-content: stretch;
  align-items: stretch;
  border-radius: 100px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  height: 36px;
  min-width: 125px;
  overflow: hidden;
  position: relative;
  cursor: pointer;
`;

const SearchSpan = styled.span`
  color: rgb(115,115,115);
  cursor: pointer;
  font-size: 16px;
  line-height: 20px;
  margin: 0;
  text-wrap: wrap;
  position: relative;

  &:before {
    content: "Search"
  }
`;

const ActionWrapper = styled.div`
  align-items: center;
  align-content: center;
  display: flex;
  flex-direction: row;
  flex-basis: 128px;
  justify-content: flex-end;
  position: relative;
  flex-grow: 1;
`;

const ActionLinkWrapper = styled.div`
  align-content: stretch;
  align-items: stretch;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  position: relative;
  margin-left: 16px;
  font-size: 14px;
`;

const LoginButton = styled(Link)`
  border-radius: 8px; 
  text-decoration: none;
  align-items: center;
  font-weight: 600;
  justify-content: center;
  text-wrap: nowrap;
  color: white; 
  background-color: rgb(0, 150, 245);
  cursor: pointer; 
  display: flex;
  height: 32px; 
  position: relative;
  text-align: center;
  padding-left: 16px; 
  padding-right: 16px; 
  flex-direction: row;

  &:hover {
    background-color: rgb(25, 120, 240);
  }
`;

const SignupButton = styled(Link)`
  border-radius: 8px; 
  text-decoration: none;
  align-items: center;
  font-weight: 600;
  justify-content: center;
  text-wrap: nowrap;
  color: rgb(0, 150, 245); 
  background-color: white;
  cursor: pointer; 
  display: flex;
  height: 32px; 
  position: relative;
  text-align: center;  
  padding-right: 16px; 
  flex-direction: row;

  &:hover {
    color: rgb(0, 50, 100);
  }  
`;

class Header extends React.Component<any, any> {
    override render() {      
      return (
        <HeaderWrapper>
          <div style={{flexBasis: "60px"}}></div>
          <div className={styles.fixedHeader}>
            <div className={styles.innerHeader}>
              <LogoWrapper>
                <Link to="/">
                  <LogoImage />
                </Link>
              </LogoWrapper>
              <SearchWrapper>         
                <SearchBox aria-label="Search input box" type="text" placeholder="" >
                </SearchBox>
                <div className={styles.searchBoxDiv1}>
                  <div className={styles.searchBoxDiv2}>
                    <div className={styles.searchBoxDiv3}>
                      <SearchImage />
                    </div>
                    <SearchSpan />                      
                  </div>                  
                </div>
              </SearchWrapper>
              <ActionWrapper>
                <div className={styles.actionWrapperInner}>
                  <ActionLinkWrapper>
                    <LoginButton to="/login">
                      Log In
                    </LoginButton>
                  </ActionLinkWrapper>
                  <ActionLinkWrapper>
                    <SignupButton to="/signup">
                      Sign Up
                    </SignupButton>
                  </ActionLinkWrapper>
                </div>
              </ActionWrapper>
            </div>
          </div>
        </HeaderWrapper>
      )
    }
  }
  
  export default Header; 