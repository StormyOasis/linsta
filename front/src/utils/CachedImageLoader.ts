import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}