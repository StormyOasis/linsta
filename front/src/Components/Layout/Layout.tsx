import React from "react";
import { Route, Routes} from "react-router-dom";
import { connect } from "react-redux";
import { Store } from "src/common/state/store";
import { LoginLayout } from "./Login/LoginLayout";
import { MainLayout } from "./Main/MainLayout";
import Header from "./Header";
import { styled } from "styled-components";
import { SignupLayout } from "./Signup/SignupLayout";

const Section = styled.section`
    display: flex;
    flex-direction: column;
    flex-grow: 1;
`;

const ModalContainer = styled.div`
    justify-content: center;  
    display: flex;
    align-items: flex-start;
    background-color: rgba(0,0,0,.6);
    min-width: 100%;
    max-width: 100%;
    z-index: 9999;
    position: fixed;
`

type LayoutProps = {
    location: string,
    isLoggedIn?: boolean
};

type LayoutState = {};

class Layout extends React.Component<LayoutProps, LayoutState> {
    
    constructor (props: LayoutProps) {
        super(props);        
    }

    renderHeader = () => {
        // Don't want to display the header on the login or signup routes
        // or if user is logged in
       /* if(this.props.isLoggedIn || 
            this.props.location.endsWith("/login") || 
            this.props.location.endsWith("/signup"))  {
                return null;
        }*/

        return (
            <Header />
        );
    }

    renderLayout = () => {
        return (  
            <>
                <Section id="mainSectionContainer">
                    {this.renderHeader()}
                    <Routes>
                        <Route path="*" element={<MainLayout />} />
                        <Route path="/login/*" element={<LoginLayout />} />
                        <Route path="/signup/*" element={<SignupLayout />} />
                    </Routes>
                </Section>
            </>          
        );
    }

    override render() {
        return (
            <>
                <ModalContainer id="modalContainer" />
                {this.renderLayout()}
            </>
        );
    }
}

const mapStateToProps = (state: Store) => ({
    isLoggedIn: state.isLoggedIn,
});

export default connect(mapStateToProps)(Layout);