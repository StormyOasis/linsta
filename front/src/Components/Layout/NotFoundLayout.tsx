import React from 'react';

import { Div, FlexColumn } from "../Common/CombinedStyling";
import StyledLink from "../Common/StyledLink";
import Theme from '../Themes/Theme';
import SideBar from './Main/SideBar';
import { historyUtils } from '../../utils/utils';
import { useAppSelector } from '../Redux/redux';
import { AuthUser } from '../../api/Auth';

const NotFoundLayout: React.FC = () => {
    const authUser:AuthUser = useAppSelector((value:any) => value?.auth?.user);
    const isLoggedIn = (!historyUtils.isServer && authUser && authUser.token != null);

    return (
        <Theme>
            <Div>
                {isLoggedIn && <SideBar />}
                <FlexColumn $height="fit-content" $width="fit-content" $alignContent="center" 
                    style={{padding: "40px", margin: "auto", flexWrap: "wrap", alignItems: "center"}}>

                    <Div $lineHeight="128px" $fontSize="24px" $fontWeight="700">Sorry, this page is not available</Div>
                    <StyledLink to="/">Go Back to Linstagram</StyledLink>
                </FlexColumn>
            </Div>
        </Theme>        
    );
};

export default NotFoundLayout;