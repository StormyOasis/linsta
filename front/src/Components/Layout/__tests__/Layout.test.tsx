import React from "react";
import { screen } from "@testing-library/react";
import Layout from "../Layout";
import '@testing-library/jest-dom';
import { renderWithStore } from '../../../utils/test-utils';

// Mocks for all layouts and components used in routes
jest.mock("../Login/LoginLayout", () => () => <div data-testid="login-layout" />);
jest.mock("../Main/MainLayout", () => () => <div data-testid="main-layout" />);
jest.mock("../Header", () => () => <div data-testid="header" />);
jest.mock("../Signup/SignupLayout", () => () => <div data-testid="signup-layout" />);
jest.mock("../Login/ForgotPasswordLayout", () => () => <div data-testid="forgot-password-layout" />);
jest.mock("../Login/ChangePasswordLayout", () => () => <div data-testid="change-password-layout" />);
jest.mock("../Main/Modals/ModalManager", () => () => <div data-testid="modal-manager" />);
jest.mock("../Profile/ProfileLayout", () => (props: any) => <div data-testid={`profile-layout${props.edit ? "-edit" : ""}`} />);
jest.mock("../NotFoundLayout", () => () => <div data-testid="not-found-layout" />);
jest.mock("../Explore/ExploreLayout", () => () => <div data-testid="explore-layout" />);
jest.mock("../../Common/Private", () => ({ children }: any) => <>{children}</>);

describe("Layout routing", () => {
    const renderWithRoute = (route: string) =>
        renderWithStore(<Layout />, { route });

    it("renders MainLayout for /", () => {
        const r = renderWithRoute("/");
        expect(screen.getByTestId("main-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders ProfileLayout in edit mode for /edit", () => {
        const r = renderWithRoute("/edit");
        expect(screen.getByTestId("profile-layout-edit")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders ExploreLayout for /explore", () => {
        const r = renderWithRoute("/explore");
        expect(screen.getByTestId("explore-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders LoginLayout for /login", () => {
        const r = renderWithRoute("/login");
        expect(screen.getByTestId("login-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders SignupLayout for /signup", () => {
        const r = renderWithRoute("/signup");
        expect(screen.getByTestId("signup-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders ForgotPasswordLayout and Header for /forgot", () => {
        const r = renderWithRoute("/forgot");
        expect(screen.getByTestId("header")).toBeInTheDocument();
        expect(screen.getByTestId("forgot-password-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders ChangePasswordLayout and Header for /change_password", () => {
        const r = renderWithRoute("/change_password");
        expect(screen.getByTestId("header")).toBeInTheDocument();
        expect(screen.getByTestId("change-password-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders NotFoundLayout for /404", () => {
        const r = renderWithRoute("/404");
        expect(screen.getByTestId("not-found-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("redirects /:userName/ to /:userName", () => {
        const r = renderWithRoute("/alice/");
        expect(screen.getByTestId("profile-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders ProfileLayout for /:userName", () => {
        const r = renderWithRoute("/bob");
        expect(screen.getByTestId("profile-layout")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("always renders ModalManager", () => {
        const r = renderWithRoute("/");
        expect(screen.getByTestId("modal-manager")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });
});