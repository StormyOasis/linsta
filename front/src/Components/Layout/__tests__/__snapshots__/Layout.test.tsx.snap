// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout routing always renders ModalManager 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="main-layout"
    />
  </div>
</div>
`;

exports[`Layout routing redirects /:userName/ to /:userName 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="profile-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders ChangePasswordLayout and Header for /change_password 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="header"
    />
    <div
      data-testid="change-password-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders ExploreLayout for /explore 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="explore-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders ForgotPasswordLayout and Header for /forgot 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="header"
    />
    <div
      data-testid="forgot-password-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders LoginLayout for /login 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="login-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders MainLayout for / 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="main-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders NotFoundLayout for /404 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="not-found-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders ProfileLayout for /:userName 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="profile-layout"
    />
  </div>
</div>
`;

exports[`Layout routing renders ProfileLayout in edit mode for /edit 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="profile-layout-edit"
    />
  </div>
</div>
`;

exports[`Layout routing renders SignupLayout for /signup 1`] = `
<div>
  <div
    class="CombinedStyling__Div-sc-c3jdxk-0 CombinedStyling__Flex-sc-c3jdxk-2 CombinedStyling__FlexColumn-sc-c3jdxk-4 escqfF ckFvxY leZDxE"
    id="mainSectionContainer"
  >
    <div
      data-testid="modal-manager"
    />
    <div
      data-testid="signup-layout"
    />
  </div>
</div>
`;
